### PR TITLE
Add Dynamoose user

### PR DESCRIPTION
**Pull request by DomBelcher**
This adds a Dynamoose User schema for User data from Alma, which includes methods to add Loans and Requests to the User, and to retrieve all User's Requests/Loans from the cache.

### CODE REVIEW

File: src/user-schema.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        .then(data => {
+          retrievedData = retrievedData.concat(data)
+        }))
+    })
+
+    return Promise.all(promises).then(() => retrievedData)
+  },
+
+  getLoanData: function (loanModel) {
+    return this.getData(loanModel, this.loan_ids, 'loan_id')
```
Comment (stephenrob) [inline src/user-schema.js]: Are you not creating instances of loan objects on the class in an array just storing the Ids ?

**Review by stephenrob — COMMENTED** [review]
Couple of queries about handling nested objects as commented above. The rest of the code looks fine though.

File: src/user-schema.js (hunk 1 above)
Comment (DomBelcher) [reply src/user-schema.js]: I've added a test in https://github.com/lulibrary/LAG-Alma-Utils/pull/17/commits/900740b266706473d68ade946dc4666acea121f0 which confirms that the result of getData (and hence getLoanData/getRequestData) is an array loan/request objects

File: .gitignore [hunk 2]
```diff
@@ -58,4 +58,7 @@ typings/
 .env
 
 # DynamoDB Local directory
-dynamodblocal
\ No newline at end of file
+dynamodblocal
+
+# vscode
+.vscode
```
Comment (stephenrob) [inline .gitignore]: Should this not be in your machines global .gitignore instead of the repository one ?

**Comment by stephenrob** [comment]
I'm still struggling to figure out the logic in the duplication that code climate raises as well. Seems like too much or not the right abstraction.

File: .gitignore (hunk 2 above)
Comment (DomBelcher) [reply .gitignore]: Probably yes. I'll remove it

**Comment by DomBelcher** [comment]
Possibly some of the methods/variables are poorly named. The `getLoanData` method populates the `loans` field on the `User` object, resolving the IDs in `loan_ids` to `Loan` objects from the supplied `loanModel`. `getRequestData` does the same but for `Requests`.

Since the `getLoanData` and `getRequestData` methods do much the same thing, the core functionality is abstracted to `getData`, which takes a dynamoose Model, the field in the User object containing the IDs to retrieve, and the name of the corresponding key for the Model. It does a series of batchGet queries on the Model and returns all the retrieved data.

I'll fix up some of the variable names to make it a bit clearer. Maybe there's a better way of doing it, but I don't think the inbuilt dynamoose `populate` method can deal with resolving arrays of values.

**Review by stephenrob — APPROVED** [review]
I feel happier with the current state of this. It looks like all the behaviour should work as intended and the tests pass. We will have to see whether any changes need to be made once it starts being using in the API Gateway work.

**Pull request merged**